File: springboot3/src/main/java/org/acme/repository/FruitRepository.java
package org.acme.repository;

import static java.util.Objects.requireNonNull;
import static org.jooq.impl.DSL.field;
import static org.jooq.impl.DSL.name;
import static org.jooq.impl.DSL.noCondition;
import static org.jooq.impl.DSL.sequence;
import static org.jooq.impl.DSL.table;
import static org.jooq.impl.DSL.val;
import static org.jooq.impl.SQLDataType.BIGINT;
import static org.jooq.impl.SQLDataType.NUMERIC;
import static org.jooq.impl.SQLDataType.VARCHAR;
import static org.springframework.transaction.annotation.Propagation.SUPPORTS;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Optional;

import org.acme.domain.Address;
import org.acme.domain.Fruit;
import org.acme.domain.Store;
import org.acme.domain.StoreFruitPrice;
import org.jooq.Condition;
import org.jooq.DSLContext;
import org.jooq.Field;
import org.jooq.Record;
import org.jooq.Result;
import org.jooq.Sequence;
import org.jooq.Table;

import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Repository
public class FruitRepository {
  private static final Table<Record> FRUITS = table(name("fruits"));
  private static final Field<Long> ID = field(name("id"), BIGINT);
  private static final Field<String> NAME = field(name("name"), VARCHAR);
  private static final Field<String> DESCRIPTION = field(name("description"), VARCHAR);
  private static final Sequence<Long> FRUITS_SEQUENCE = sequence(name("fruits_seq"), BIGINT);

  private static final Table<Record> F = FRUITS.as("f");
  private static final Table<Record> SFP = table(name("store_fruit_prices")).as("sfp");
  private static final Table<Record> S = table(name("stores")).as("s");

  private static final Field<Long> FRUIT_ID = field(name("f", "id"), BIGINT);
  private static final Field<String> FRUIT_NAME = field(name("f", "name"), VARCHAR);
  private static final Field<String> FRUIT_DESCRIPTION = field(name("f", "description"), VARCHAR);
  private static final Field<Long> PRICE_FRUIT_ID = field(name("sfp", "fruit_id"), BIGINT);
  private static final Field<Long> PRICE_STORE_ID = field(name("sfp", "store_id"), BIGINT);
  private static final Field<BigDecimal> PRICE = field(name("sfp", "price"), NUMERIC);
  private static final Field<Long> STORE_ID = field(name("s", "id"), BIGINT);
  private static final Field<String> STORE_NAME = field(name("s", "name"), VARCHAR);
  private static final Field<String> STORE_CURRENCY = field(name("s", "currency"), VARCHAR);
  private static final Field<String> STORE_ADDRESS = field(name("s", "address"), VARCHAR);
  private static final Field<String> STORE_CITY = field(name("s", "city"), VARCHAR);
  private static final Field<String> STORE_COUNTRY = field(name("s", "country"), VARCHAR);

  private final DSLContext dsl;

  public FruitRepository(DSLContext dsl) {
    this.dsl = dsl;
  }

  @Transactional(propagation = SUPPORTS, readOnly = true)
  public Optional<Fruit> findByName(String name) {
    return fetchFruits(FRUIT_NAME.eq(name)).stream().findFirst();
  }

  @Transactional(propagation = SUPPORTS, readOnly = true)
  public List<Fruit> findAll() {
    return fetchFruits(noCondition());
  }

  public Fruit save(Fruit fruit) {
    Field<Long> id = fruit.getId() == null ? FRUITS_SEQUENCE.nextval() : val(fruit.getId());
    Long savedId = this.dsl.insertInto(FRUITS)
        .set(ID, id)
        .set(NAME, fruit.getName())
        .set(DESCRIPTION, fruit.getDescription())
        .returningResult(ID)
        .fetchOne(ID);

    fruit.setId(requireNonNull(savedId, "Insert did not return a fruit id"));
    return fruit;
  }

  private List<Fruit> fetchFruits(Condition condition) {
    var rows = this.dsl
        .select(
            FRUIT_ID,
            FRUIT_NAME,
            FRUIT_DESCRIPTION,
            STORE_ID,
            STORE_NAME,
            STORE_CURRENCY,
            STORE_ADDRESS,
            STORE_CITY,
            STORE_COUNTRY,
            PRICE)
        .from(F)
        .leftJoin(SFP).on(PRICE_FRUIT_ID.eq(FRUIT_ID))
        .leftJoin(S).on(STORE_ID.eq(PRICE_STORE_ID))
        .where(condition)
        .orderBy(FRUIT_ID, STORE_ID)
        .fetch();

    return mapRows(rows);
  }

  private static List<Fruit> mapRows(Result<? extends Record> rows) {
    var fruits = new LinkedHashMap<Long, Fruit>();

    for (var row : rows) {
      Long fruitId = row.get(FRUIT_ID);
      Fruit fruit = fruits.computeIfAbsent(fruitId, ignored -> {
        var mappedFruit = new Fruit(fruitId, row.get(FRUIT_NAME), row.get(FRUIT_DESCRIPTION));
        mappedFruit.setStorePrices(new ArrayList<>());
        return mappedFruit;
      });

      Long storeId = row.get(STORE_ID);
      if (storeId != null) {
        var address = new Address(
            row.get(STORE_ADDRESS),
            row.get(STORE_CITY),
            row.get(STORE_COUNTRY));
        var store = new Store(
            storeId,
            row.get(STORE_NAME),
            address,
            row.get(STORE_CURRENCY));
        fruit.getStorePrices().add(new StoreFruitPrice(store, fruit, row.get(PRICE)));
      }
    }

    return List.copyOf(fruits.values());
  }
}

File: aspnet10/src/Repository/FruitRepository.cs
using AspNet10.Domain;
using Dapper;
using Npgsql;

namespace AspNet10.Repository;

public sealed class FruitRepository(NpgsqlDataSource dataSource) : IFruitRepository
{
    private const string SelectSql = """
        SELECT f.id          AS "FruitId",
               f.name        AS "FruitName",
               f.description AS "FruitDescription",
               s.id          AS "StoreId",
               s.name        AS "StoreName",
               s.currency    AS "StoreCurrency",
               s.address     AS "StoreAddress",
               s.city        AS "StoreCity",
               s.country     AS "StoreCountry",
               sfp.price     AS "Price"
          FROM fruits f
          LEFT JOIN store_fruit_prices sfp ON sfp.fruit_id = f.id
          LEFT JOIN stores s ON s.id = sfp.store_id
        """;

    private const string InsertSql = """
        INSERT INTO fruits (id, name, description)
        VALUES (nextval('fruits_seq'), @Name, @Description)
        RETURNING id
        """;

    public async Task<IReadOnlyList<Fruit>> FindAllAsync(
        CancellationToken cancellationToken = default)
    {
        const string sql = SelectSql + " ORDER BY f.id, s.id";
        var rows = await QueryAsync(sql, null, cancellationToken);
        return MapRows(rows);
    }

    public async Task<Fruit?> FindByNameAsync(
        string name,
        CancellationToken cancellationToken = default)
    {
        const string sql = SelectSql + " WHERE f.name = @Name ORDER BY f.id, s.id";
        var rows = await QueryAsync(sql, new { Name = name }, cancellationToken);
        return MapRows(rows).SingleOrDefault();
    }

    public async Task<Fruit> SaveAsync(
        Fruit fruit,
        CancellationToken cancellationToken = default)
    {
        await using var connection = await dataSource.OpenConnectionAsync(cancellationToken);
        await using var transaction = await connection.BeginTransactionAsync(cancellationToken);

        var command = new CommandDefinition(
            InsertSql,
            new { fruit.Name, fruit.Description },
            transaction,
            cancellationToken: cancellationToken);
        fruit.Id = await connection.QuerySingleAsync<long>(command);

        await transaction.CommitAsync(cancellationToken);
        return fruit;
    }

    private async Task<IEnumerable<FruitRow>> QueryAsync(
        string sql,
        object? parameters,
        CancellationToken cancellationToken)
    {
        await using var connection = await dataSource.OpenConnectionAsync(cancellationToken);
        return await connection.QueryAsync<FruitRow>(
            new CommandDefinition(sql, parameters, cancellationToken: cancellationToken));
    }

    private static IReadOnlyList<Fruit> MapRows(IEnumerable<FruitRow> rows)
    {
        var fruits = new Dictionary<long, Fruit>();

        foreach (var row in rows)
        {
            if (!fruits.TryGetValue(row.FruitId, out var fruit))
            {
                fruit = new Fruit
                {
                    Id = row.FruitId,
                    Name = row.FruitName,
                    Description = row.FruitDescription
                };
                fruits.Add(row.FruitId, fruit);
            }

            if (row.StoreId is { } storeId)
            {
                var store = new Store
                {
                    Id = storeId,
                    Name = row.StoreName!,
                    Currency = row.StoreCurrency!,
                    Address = new Address(row.StoreAddress!, row.StoreCity!, row.StoreCountry!)
                };
                fruit.StorePrices.Add(new StoreFruitPrice
                {
                    Id = new StoreFruitPriceId(storeId, row.FruitId),
                    Store = store,
                    Fruit = fruit,
                    Price = row.Price!.Value
                });
            }
        }

        return fruits.Values.ToArray();
    }

    private sealed class FruitRow
    {
        public long FruitId { get; init; }
        public required string FruitName { get; init; }
        public string? FruitDescription { get; init; }
        public long? StoreId { get; init; }
        public string? StoreName { get; init; }
        public string? StoreCurrency { get; init; }
        public string? StoreAddress { get; init; }
        public string? StoreCity { get; init; }
        public string? StoreCountry { get; init; }
        public decimal? Price { get; init; }
    }
}
